### DISCUSSION TITLE
Constraining constants to be real for imaginary-value data

### DISCUSSION THREAD

**Discussion by Elk113** [MilesCranmer/PySR#768] [Q&A] [answered]
Thanks for developing pySR. I'm really enjoying it.
I'm stuck with the following question: 
How can I enforce a customized unary operator on all leaf nodes (terminmal nodes)? But I don't want to allow it on inner nodes.
Can you please advise me?
Thanks

**Comment by MilesCranmer** [comment]
You could do `constraints={"myoperator": 1}`? See https://ai.damtp.cam.ac.uk/pysr/api/#working-with-complexities

**Reply by Elk113** [reply]
Thanks for your respond. I tried this, but it doesn't make sure that the customized unary function is applied to EACH terminal node.

**Reply by MilesCranmer** [reply]
Oh, in that case I'm not sure I follow. You want all constants and variables to pass through a given operator? Why not scale the variables before inputting to the search?

**Reply by Elk113** [reply]
As I work with impedance operators the binary operators will result in a complex number. The input values are real though. So my tree works with complex numbers, but the terminal nodes are real numbers. So my struggle is the switch between real and complex numbers which I hoped to resolve with an unary operator on all terminal nodes. I hope this makes sense to you.

**Reply by MilesCranmer** [reply]
Oh, I see! That makes much more sense. Thanks. Will write new comment –

**Comment by MilesCranmer** [comment, marked as answer]
So you can either do it with a template expression or a custom loss function. I suppose I would try a custom loss first? Then you won't need to do any manual postprocessing as the raw expression would have nearly-real constants.

```julia
function my_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    prediction, flag = eval_tree_array(tree, dataset.X, options)
    if !flag
        return L(Inf)
    end

    square_error = sum(i -> abs2(prediction[i] - dataset.y[i]), eachindex(prediction))
    mse = square_error / length(dataset.y)

    imaginary_value_penalty = sum(
        node -> if node.degree == 0 && node.constant
            val = node.val
            penalty = abs(imag(val) / (abs(val) + 1e-10))
            L(penalty)  # penalize any non-real part of constants
        else
            zero(L)
        end,
        tree
    )
    
    return mse + 1000 * imaginary_value_penalty
end
```

(The `L` is just the numeric type of the loss - which in this case is probably `Float32`)

So while the numerical constants will still have _some_ imaginary component, the search will be encouraged to minimise it. It's a bit easier than needing to convert the constants. It also means you will have sympy compatibility.


You can pass this to PySR like this:

```python
PySRRegressor(
    ... # other args
    loss_function="""
        function my_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
            prediction, flag = eval_tree_array(tree, dataset.X, options)
            if !flag
                return L(Inf)
            end
        
            square_error = sum(i -> abs2(prediction[i] - dataset.y[i]), eachindex(prediction))
            mse = square_error / length(dataset.y)
        
            imaginary_value_penalty = sum(
                node -> if node.degree == 0 && node.constant
                    val = node.val
                    penalty = abs(imag(val) / (abs(val) + 1e-10))
                    L(penalty)
                else
                    zero(L)
                end,
                tree
            )
            
            return mse + 1000 * imaginary_value_penalty
        end"""
)
```

**Reply by MilesCranmer** [reply]
@Elk113 and you could also make the imaginary value penalty larger if needed by multiplying it with a large constant?

**Reply by Elk113** [reply]
Thanks a lot for this quick and elaborated answer. I'm looking forward to testing it!

**Reply by MilesCranmer** [reply]
Did it work?

**Reply by Elk113** [reply]
Yes, it did! Thank you very much for your help!
